### PR TITLE
New fake azure client

### PR DESCRIPTION
**Pull request by asalkeld**
```release-note
NONE
```

The point of this is to have a deeper style of integration test that allows us to test the whole plugin with just azure and kube client fakes. We can then have greater confidence in what actions cause node rotations.
/cc @jim-minter

### CODE REVIEW

File: pkg/util/azureclient/fake/compute.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+// Client Fakes base method
+func (s *FakeVirtualMachineScaleSetsClient) Client() autorest.Client {
+	return allwaysDoneClient()
+}
+
+func (s *FakeVirtualMachineScaleSetsClient) scale(ctx context.Context, resourceGroupName string, ss *compute.VirtualMachineScaleSet) error {
+	var have int64
+	for _, vm := range s.az.Vms {
+		if *ss.Name == *vm.Tags["scaleset"] {
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/compute.go]: don't do this ;)

File: pkg/util/azureclient/fake/compute.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+			have++
+		}
+	}
+	s.az.log.Debugf("scale have:%d, cap:%d", have, *ss.Sku.Capacity)
+	if have > *ss.Sku.Capacity {
+		return fmt.Errorf("should not be automatically scaling down")
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/compute.go]: we'll need this soon surely

File: pkg/util/azureclient/fake/compute.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+			VirtualMachineScaleSetVMProperties: &compute.VirtualMachineScaleSetVMProperties{
+				OsProfile: &compute.OSProfile{ComputerName: &compName},
+			},
+			Location: ss.Location,
+			Plan:     ss.Plan,
+			Tags:     tags,
+			Zones:    ss.Zones,
+		}
+		s.az.Vms = append(s.az.Vms, vm)
+		s.az.VirtualMachineScaleSetVMsClient.Start(ctx, resourceGroupName, *ss.Name, *vm.InstanceID)
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/compute.go]: do we really want to do this?

File: pkg/util/azureclient/fake/compute.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	s.scale(ctx, resourceGroupName, &parameters)
+	return nil
+}
+
+// Delete Fakes base method
+func (s *FakeVirtualMachineScaleSetsClient) Delete(ctx context.Context, resourceGroupName, VMScaleSetName string) error {
+	for i, ss := range s.az.Ssc {
+		if VMScaleSetName == *ss.Name {
+			*ss.Sku.Capacity = 0
+			s.scale(ctx, resourceGroupName, &ss)
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/compute.go]: would currently return not implemented but you're not checking the error?

File: pkg/util/azureclient/fake/compute.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+			return nil
+		}
+	}
+	return nil
+}
+
+// UpdateInstances Fakes base method
+func (s *FakeVirtualMachineScaleSetsClient) UpdateInstances(ctx context.Context, resourceGroupName, VMScaleSetName string, VMInstanceIDs compute.VirtualMachineScaleSetVMInstanceRequiredIDs) error {
+	// not implemented yet
+	return nil
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/compute.go]: don't know if this should return a not implemented error

File: pkg/util/azureclient/fake/deployments.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+				return
+			}
+			err = sa.UnmarshalJSON(sab)
+			if err != nil {
+				result.Future.Response().StatusCode = 500
+				return
+			}
+
+			updated := false
+			for a, acct := range d.az.Accts {
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/deployments.go]: should be calling a CreateOrUpdate function

File: pkg/util/azureclient/fake/deployments.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+					d.az.Accts[a] = sa
+					updated = true
+				}
+			}
+			if !updated {
+				d.az.Accts = append(d.az.Accts, sa)
+			}
+		} else if strings.Contains(rMap["type"].(string), "Microsoft.Compute/virtualMachineScaleSets") {
+			vm := compute.VirtualMachineScaleSet{}
+			rb, _ := json.Marshal(rMap)
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/deployments.go]: make a helper function and check errors

File: pkg/util/azureclient/fake/deployments.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+			vm.UnmarshalJSON(rb)
+			d.az.VirtualMachineScaleSetsClient.CreateOrUpdate(ctx, resourceGroupName, *vm.Name, vm)
+		}
+	}
+
+	// store in memory the resources that are created so that other api requests can work with them
+	err = nil
```
Comment (jim-minter) [inline pkg/util/azureclient/fake/deployments.go]: dodgy!

**Comment by jim-minter** [comment]
/retest
/lgtm

@asalkeld please deal with the comments in a follow-up

File: pkg/cluster/kubeclient/roundtrip.go [hunk 9]
```diff
@@ -167,5 +169,12 @@ func NewKubeclient(log *logrus.Entry, config *v1.Config, disableKeepAlives bool)
 		client: cli,
 		seccli: seccli,
 	}, nil
+}
 
+func NewFakeKubeclient(log *logrus.Entry, cli *fake.Clientset, seccli *fakesec.Clientset) Kubeclient {
+	return &kubeclient{
+		log:    log,
+		client: cli,
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/cluster/kubeclient/roundtrip.go]: ok

File: pkg/cluster/types.go [hunk 10]
```diff
@@ -136,6 +138,47 @@ func NewSimpleUpgrader(ctx context.Context, log *logrus.Entry, cs *api.OpenShift
...
+	arm, err := arm.New(ctx, log, cs, testConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	u := &simpleUpgrader{
+		Kubeclient: kubeclient,
+
+		testConfig:     testConfig,
+		accountsClient: azs.AccountsClient,
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/cluster/types.go]: ok

File: pkg/plugin/integration_test.go [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+
+func (w fakeResponseWrapper) Stream() (io.ReadCloser, error) {
+	return nil, nil
+}
+
+func newFakeResponseWrapper(raw []byte) fakeResponseWrapper {
+	return fakeResponseWrapper{raw: raw}
+}
+
+type fakeResponseWrapper struct {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/integration_test.go]: ok

File: pkg/plugin/integration_test.go [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+
+func (w fakeResponseWrapper) DoRaw() ([]byte, error) {
+	return w.raw, nil
+}
+
+func (w fakeResponseWrapper) Stream() (io.ReadCloser, error) {
+	return nil, nil
+}
+
+func newFakeResponseWrapper(raw []byte) fakeResponseWrapper {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/integration_test.go]: ok

File: pkg/plugin/integration_test.go [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+			Kind: "securitycontextconstraints",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "privileged",
+		},
+		AllowPrivilegedContainer: true,
+		Users:                    []string{"system:admin"},
+	}
+	seccli := fakesec.NewSimpleClientset()
+	seccli.PrependReactor("get", "*", func(action k8stesting.Action) (bool, runtime.Object, error) {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/integration_test.go]: ok

File: pkg/plugin/integration_test.go [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+			},
+		}
+		return true, pod, nil
+	})
+	cli.PrependReactor("get", "nodes", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		get := action.(k8stesting.GetAction)
+		node := &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      get.GetName(),
+				Namespace: get.GetNamespace(),
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/integration_test.go]: ok

File: pkg/util/azureclient/fake/deployments.go (hunk 6 above)
Comment (asalkeld) [reply pkg/util/azureclient/fake/deployments.go]: I couldn't find a CreateOrUpdate only Create and Update

File: pkg/util/azureclient/fake/compute.go (hunk 5 above)
Comment (asalkeld) [reply pkg/util/azureclient/fake/compute.go]: good idea

File: pkg/util/azureclient/fake/deployments.go (hunk 8 above)
Comment (asalkeld) [reply pkg/util/azureclient/fake/deployments.go]: cleaned up

File: pkg/plugin/integration_test.go [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+	seccli := fakesec.NewSimpleClientset()
+	seccli.PrependReactor("get", "*", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		get := action.(k8stesting.GetAction)
+		if get.GetResource().Resource == "securitycontextconstraints" {
+			return true, &priv, nil
+		}
+		return false, nil, fmt.Errorf("does not exist")
+	})
+	seccli.PrependReactor("update", "*", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		get := action.(k8stesting.UpdateAction)
```
Comment (charlesakalugwu) [inline pkg/plugin/integration_test.go]:
```suggestion
		update := action.(k8stesting.UpdateAction)
```

File: pkg/plugin/integration_test.go [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+		if get.GetResource().Resource == "securitycontextconstraints" {
```
Comment (charlesakalugwu) [inline pkg/plugin/integration_test.go]:
```suggestion
			if update.GetResource().Resource == "securitycontextconstraints" {
```

File: pkg/plugin/integration_test.go [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+	updateBlobService := updateblob.NewBlobService(bsc)
+	beforeBlob, err := updateBlobService.Read()
+	// - update
+	if err := p.CreateOrUpdate(ctx, cs, true, getFakeDeployer(log, cs, az)); err != nil {
+		t.Errorf("plugin.CreateOrUpdate [update] error = %v", err)
+	}
+	// - assert that the hashes are the same
+	afterBlob, err := updateBlobService.Read()
+	if !reflect.DeepEqual(beforeBlob, afterBlob) {
+		t.Errorf("unexpected result:\n%#v\nexpected:\n%#v", spew.Sprint(beforeBlob), spew.Sprint(afterBlob))
```
Comment (charlesakalugwu) [inline pkg/plugin/integration_test.go]: maybe use `deep.Equal` to show the differences instead? Something like...

```suggestion
		t.Errorf("unexpected blob difference: \n %#v", deep.Equal(beforeBlob, afterBlob))
```

File: pkg/plugin/integration_test.go [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+			return true, &priv, nil
+		}
+		return false, nil, fmt.Errorf("does not exist")
+	})
+	seccli.PrependReactor("update", "*", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		get := action.(k8stesting.UpdateAction)
+		if get.GetResource().Resource == "securitycontextconstraints" {
+			return true, &priv, nil
+		}
+		return false, nil, fmt.Errorf("does not exist")
```
Comment (charlesakalugwu) [inline pkg/plugin/integration_test.go]: can we improve this error message? I am not sure about what doesn't exist here

File: pkg/plugin/integration_test.go [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+		AllowPrivilegedContainer: true,
+		Users:                    []string{"system:admin"},
+	}
+	seccli := fakesec.NewSimpleClientset()
+	seccli.PrependReactor("get", "*", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		get := action.(k8stesting.GetAction)
+		if get.GetResource().Resource == "securitycontextconstraints" {
+			return true, &priv, nil
+		}
+		return false, nil, fmt.Errorf("does not exist")
```
Comment (charlesakalugwu) [inline pkg/plugin/integration_test.go]: can we improve this error message? I am not sure about what doesn't exist here

File: pkg/plugin/integration_test.go [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+	return nil
+}
+
+func (w fakeResponseWrapper) DoRaw() ([]byte, error) {
+	return w.raw, nil
+}
+
+func (w fakeResponseWrapper) Stream() (io.ReadCloser, error) {
+	return nil, nil
+}
```
Comment (charlesakalugwu) [inline pkg/plugin/integration_test.go]: what interface are `DoRaw` and `Stream` from?

**Pull request merged**